samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreRestoreManager.java

```diff
@@ -254,8 +263,10 @@ static CompletableFuture<Void> restoreStores(String jobName, String jobId, TaskN
         throw new SamzaException(String.format("Error deleting store directory: %s", storeDir), e);
       }
 
+      // If getDeletedBlobs is enabled - always restore so that we get all the blobs, including the deleted blobs,
+      // immediately restore it locally and backup to create new checkpoint.
       boolean shouldRestore = shouldRestore(taskName.getTaskName(), storeName, dirIndex,
-          storeCheckpointDir, storageConfig, dirDiffUtil);
+          storeCheckpointDir, storageConfig, dirDiffUtil) || getDeletedBlobs;
```

Comment:
> Any scenarios where shouldRestore returns false but we want to force restore?
Yes.
When there is no diff between remote and local snapshot, shouldRestore could return false. However, if the SnapshotIndex was deleted, we want to force restore everything (and then backup, create new checkpoint etc). This is because init() delegates this to restore and only gets the snapshotIndex with getDeleted. It does not actually create a new copy of everything.

Reply:
If shouldRestore returned false during first restore, there would have been no restore, no failures, and we would not be restoring with getDeleted = true. 

E.g. if the last checkpoint directory was already present and was identical to remote snapshot, it would be copied over to the store directory and restore would be skipped. 

It seems like this would be an issue, since the files could actually be deleted in remote store, but restore manager wouldn't realize it, and backup manager would only upload the delta based on the latest snapshot index. So the files could eventually get compacted.

So do we need to remove the shouldRestore check and always force a restore upon startup to detect correctly if we need to reupload?

Can sync up offline if I'm misunderstanding something.

Reply:
Please let me know if this sounds okay after our discussion.

Reply:
Yes, makes sense. To close the loop:
Snapshot index blob is deleted after all no-longer-needed file blobs in it are deleted.
We are using the failure to get snapshot index in init with getDeletedBlob = false as the signal that some of its contained blobs are missing and the entire store needs to be restored. This is also why the higher precedence for getDeletedBlobs is required.
For the rare race condition where file blobs were deleted but snapshot index wasn't yet, restore manager init will work, restore will work, but a subsequent commit will fail and restart the container. By that time, the SnapshotIndex should be deleted by the orphaned container.